ledger/src/blockstore.rs

```diff
@@ -1702,6 +1730,46 @@ impl Blockstore {
         })
     }
 
+    fn get_any_valid_slot_entries(&self, slot: Slot, start_index: u64) -> Vec<Entry> {
+        // lowest_cleanup_slot is the last slot that was not cleaned up by
+        // LedgerCleanupService
+        let lowest_cleanup_slot = self.lowest_cleanup_slot.read().unwrap();
+        if *lowest_cleanup_slot > slot {
+            return vec![];
+        }
+
+        let slot_meta_cf = self.db.column::<cf::SlotMeta>();
+        let slot_meta = slot_meta_cf.get(slot).unwrap_or_default();
+        if slot_meta.is_none() {
+            return vec![];
+        }
+
+        let slot_meta = slot_meta.unwrap();
+        // Find all the ranges for the completed data blocks
+        let completed_ranges = Self::get_completed_data_ranges(
+            start_index as u32,
+            &slot_meta.completed_data_indexes[..],
+            slot_meta.consumed as u32,
+        );
+        if completed_ranges.is_empty() {
+            return vec![];
+        }
```

Comment:
I thihnk the beginning of this function to here can be extracted into a `get_completed_ranges() -> Result<Vec<Entry>>` function where:

the error condition is:

`if *lowest_cleanup_slot > slot {
        return Err(BlockstoreError::SlotCleanedUp);
  }`

Reply:
Do you mean `Result<Vec<(u32, u32)>>` for the return type?

Reply:
ah yup!